### PR TITLE
BUG: fillna overwriting other missing vals

### LINKED ISSUE #40498: BUG: Series.fillna is altering None values of unspecified columns
- [x] I have checked that this issue has not already been reported.
- [x] I have confirmed this bug exists on the latest version of pandas (1.0.5)

`series = pd.Series({'1':1,'2':None,'3':np.nan, '4':''}) `

will create a series which looks like:

    1       1
    2    None
    3     NaN
    4        
    dtype: object

When using the fillna method on this series on index 3 it changes also the entry in index 2!

    series.fillna({'3':0})

This will result in a output as follows:

    1      1
    2    NaN
    3      0
    4       
    dtype: object

**Problem:**
The value at index 2 changed from None to NaN.

**Expected Behaviour:**
The value at index 2 should never be touched when using fillna and specifying the indices you want to alter.

### PR DESCRIPTION
**Pull request by mzeitlin11**
- [x] closes #40498
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

### CODE REVIEW

**Review by jreback — CHANGES_REQUESTED** [review]
i am not sure i buy that this is a bug at all. 

```
In [149]: series.isna()                                                                                                                                               
Out[149]: 
1    False
2     True
3     True
4    False
dtype: bool
```
Trying to use mixed null values in an `object` series is just not supported.

**Comment by mzeitlin11** [comment]
> i am not sure i buy that this is a bug at all.
> 
> ```
> In [149]: series.isna()                                                                                                                                               
> Out[149]: 
> 1    False
> 2     True
> 3     True
> 4    False
> dtype: bool
> ```
> 
> Trying to use mixed null values in an `object` series is just not supported.

I think the confusion from #40498 is not that `isna(None) = True`, but that the value at index 2 is modified even though the argument `{'3':0}` should ensure that only index 3 is modified.

**Comment by jreback** [comment]
> > i am not sure i buy that this is a bug at all.
> > ```
> > In [149]: series.isna()                                                                                                                                               
> > Out[149]: 
> > 1    False
> > 2     True
> > 3     True
> > 4    False
> > dtype: bool
> > ```
> > 
> > 
> > Trying to use mixed null values in an `object` series is just not supported.
> 
> I think the confusion from #40498 is not that `isna(None) = True`, but that the value at index 2 is modified even though the argument `{'3':0}` should ensure that only index 3 is modified.

I am ok with that, but your patch is not structured that way. it shouldn't be testing for nulls

**Comment by mzeitlin11** [comment]
Sorry, can you please clarify what you mean by that? 

What the patch should be doing is fixing the following: for the series `ser = Series({'1': 1,'2': None,'3': np.nan, '4': 1})`, when the fill argument '{'3':0}' hits the line I changed, it previously became `[NaN, NaN, 0, NaN]`. 

When this hits `ObjectBlock.fillna`, mask is determined from `isna(ser)`, giving `[False, True, True, False]`. So then when `putmask_inplace` is called with this mask, `None` is replaced with `np.nan` since the mask is `True` at the position.

To avoid having to change the behavior of lower level `fillna` functionality, this patch changes the fill argument to instead become `[NaN, None, 0, NaN]`, so that `None` just ends up being replaced by itself.

File: pandas/core/generic.py [hunk 1]
```diff
@@ -6524,6 +6524,11 @@ def fillna(
                         value, dtype_if_empty=object
                     )
                     value = value.reindex(self.index, copy=False)
+
```
Comment (jreback) [inline pandas/core/generic.py:6527]: i think you are better off calling .align here. What you have on L6531 is too specific, fragile and trying to fix something after the fact.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/generic.py (hunk 1 above)
Comment (mzeitlin11) [reply pandas/core/generic.py:6527]: Yeah I see what you mean. Not seeing a clear solution with .align either - will have to have some sort of ugly isna check at some point I think to support this. Will close for now and reopen if I can find a better way to do this

**Pull request closed without merging**